project/BlenderWang9487/README.rst

```diff
@@ -0,0 +1,106 @@
+========================
+DummyML
+========================
+
+
+
+Basic Information
+=================
+
+A dummy ML library for NSD course term project.
+
+Here is my Github repository:
+`DummyML <https://github.com/BlenderWang9487/DummyML.git>`_
+
+Problem to Solve
+================
+
+I want to build a machine learning library based on some mathematical theories.
+
+There are many easy-to-use hybrid frameworks that built for deep learning, 
+but fewer for other traditional ML algorithms. So I'm going to build one.
+
+Some regression, classification and clustering models (like Gaussian linear 
+regression, naive Bayes classifier, Kmeans, etc.) will be provided by this
+library. (There will be atleast three approaches for each mission. One of them
+is a "dummy" approach which can be treated as the baseline.)
+
+Prospective Users
+=================
+
+People who are interested in machine learning and want to try some different 
+ML algorithms may use my library.
+
+Data scientists can use this library's models as baselines against theirs, too.
+
+System Architecture
+===================
+
+There may or may not be a class to handle input/output data (I might just use
```

Comment:
Numpy ndarray is owned by Python.  You need to figure out a way to ensure the lifecycle when your C++ code is operating on the buffers.  A common practice is to own the data by C++ and "share" it with Python through ndarray.

You should try it yourself.  It's straight-forward in description but not so obvious in code.

Reply:
I use this [function](https://github.com/BlenderWang9487/DummyML/blob/main/include/dummyml/model.hpp#L20-L33) which let numpy refer to c++'s pointer  to handle the return values (outputs) of my APIs. Is this a way to do it?

I still need to figure it out how to handle the inputs though.

Reply:
When Python calls the line https://github.com/BlenderWang9487/DummyML/blob/fa0e29dc747dd58f45ac7b95186b1f88e9d42a2e/include/dummyml/model.hpp#L24

with the pointer `arr` already freed elsewhere, you get a crash with double-free.  Since `arr` is a raw pointer, the `PyCapsule` cannot know whether or not it is a valid pointer.  It is why when we deal with ownership in C++ we do not use a raw pointer.

You may find the use of a raw pointer for ownership in C.  That is because C does not have other constructor for pointers.

See https://yyc.solvcon.net/en/latest/nsd/07mem/mem.html#never-access-freed-memory for double free, and see https://yyc.solvcon.net/en/latest/nsd/08smart/smart.html for how we handle ownership in C++.